### PR TITLE
fix(versioning/semver-coerced): fix bugs in semver-coerced

### LINKED ISSUE #16867: semverCoerced invalid version vX.Y error
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

_No response_

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

using the `renovate.json` with custom versioning like below:
```
{
	"dependencyDashboard": true,
	"packageRules": [
		{
			"matchPackageNames": [
				"docker.io/nabeels/alp-v2",
				"docker.io/nabeels/alp-v3"
			],
			"versioning": "semver-coerced"
		}
	]
}
```

results in an error:
```
TypeError: Invalid Version: v3.15
    at new SemVer (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/semver/classes/semver.js:38:13)
    at Object.patch (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/semver/functions/patch.js:2:29)
    at Object.getPatch (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/versioning/semver-coerced/index.js:28:29)
    at /home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/filter.js:93:59
    at Array.filter (<anonymous>)
    at filterVersions (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/filter.js:89:29)
    at lookupUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/index.js:175:64)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async fetchDepUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/fetch.js:51:25)
    at async /home/ubuntu/renovateapp/node_modules/p-map/index.js:57:22"
```

minimal reproduction:
https://github.com/nabeelys/renovate-bug-docker-version-precision

### Relevant debug logs

<details><summary>Logs</summary>

```
ERROR: lookupUpdates error(packageFile="chart/values.yaml")
{
  "currentValue": "v3.15",
  "datasource": "docker",
  "depName": "docker.io/nabeels/alp-v3",
  "followTag": null,
  "pinDigests": false,
  "rollbackPrs": false,
  "updatePinnedDependencies": true,
  "unconstrainedValue": false,
  "err": {
    "message": "Invalid Version: v3.15",
    "stack": "TypeError: Invalid Version: v3.15\n    at new SemVer (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/semver/classes/semver.js:38:13)\n    at Object.patch (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/semver/functions/patch.js:2:29)\n    at Object.getPatch (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/versioning/semver-coerced/index.js:28:29)\n    at /home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/filter.js:93:59\n    at Array.filter (<anonymous>)\n    at filterVersions (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/filter.js:89:29)\n    at lookupUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/index.js:175:64)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async fetchDepUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/fetch.js:51:25)\n    at async /home/ubuntu/renovateapp/node_modules/p-map/index.js:57:22"
  }
}
ERROR: lookupUpdates error(packageFile="chart2/values.yaml")
{
  "currentValue": "3.15",
  "datasource": "docker",
  "depName": "docker.io/nabeels/alp-v2",
  "followTag": null,
  "pinDigests": false,
  "rollbackPrs": false,
  "updatePinnedDependencies": true,
  "unconstrainedValue": false,
  "err": {
    "message": "Invalid Version: 3.15",
    "stack": "TypeError: Invalid Version: 3.15\n    at new SemVer (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/semver/classes/semver.js:38:13)\n    at Object.patch (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/semver/functions/patch.js:2:29)\n    at Object.getPatch (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/versioning/semver-coerced/index.js:28:29)\n    at /home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/filter.js:93:59\n    at Array.filter (<anonymous>)\n    at filterVersions (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/filter.js:89:29)\n    at lookupUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/index.js:175:64)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async fetchDepUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/fetch.js:51:25)\n    at async /home/ubuntu/renovateapp/node_modules/p-map/index.js:57:22"
  }
}
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by nabeelsaabna**
## Changes

1. fix exception from `getPatch()` when input is not a valid semver
2. improve `isStable()` to resolve `v1.2`, `1.2`, `v1` and `1` as stable versions
 
## Context

Closes #16867 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by nabeelsaabna** [comment]
checking if documentation needs update

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/versioning/semver-coerced/index.ts [hunk 1]
```diff
@@ -9,7 +9,25 @@ export const displayName = 'Coerced Semantic Versioning';
...
+  if (!m?.groups) {
+    return false;
+  }
+
+  const major = m.groups['major'];
+  const newMinor = m.groups['minor'] ? m.groups['minor'] : '.0';
+  const newPatch = m.groups['patch'] ? m.groups['patch'] : '.0';
+  const others = m.groups['others'] ? m.groups['others'] : '';
+  const fixed = major + newMinor + newPatch + others;
+  return stable.is(fixed);
```
[the remark this answers is not in the document]
Comment (nabeelsaabna) [reply lib/modules/versioning/semver-coerced/index.ts:29]: running coerce for `v1.2-alpha` will result with `1.2.0`
that's why I'm adding missing minor and patch parts while keeping the "others" as is

File: lib/modules/versioning/semver-coerced/index.spec.ts [hunk 2]
```diff
@@ -125,6 +141,50 @@ describe('modules/versioning/semver-coerced/index', () => {
...
+      expect(semverCoerced.isStable('1.0.0-rc2')).toBeFalse();
+    });
+
+    it('should return false for a RC version', () => {
+      expect(semverCoerced.isStable('v1.0.0-rc2')).toBeFalse();
+    });
+
+    it('should return true for a short prerelease version', () => {
+      expect(semverCoerced.isStable('1.0-Beta.0')).toBeFalse();
+    });
```
[the remark this answers is not in the document]
Comment (nabeelsaabna) [reply lib/modules/versioning/semver-coerced/index.spec.ts]: much easier to read and maintain 👍

File: lib/modules/versioning/semver-coerced/index.ts [hunk 3]
```diff
@@ -9,7 +9,25 @@ export const displayName = 'Coerced Semantic Versioning';
...
+  const m = regx.exec(version);
+
+  if (!m?.groups) {
+    return false;
+  }
+
+  const major = m.groups['major'];
+  const newMinor = m.groups['minor'] ? m.groups['minor'] : '.0';
+  const newPatch = m.groups['patch'] ? m.groups['patch'] : '.0';
+  const others = m.groups['others'] ? m.groups['others'] : '';
```
Comment (JamieMagee) [inline lib/modules/versioning/semver-coerced/index.ts:27]:
```suggestion
  const newMinor = m.groups['minor'] ?? '.0';
  const newPatch = m.groups['patch'] ?? '.0';
  const others = m.groups['others'] ?? '';
```

**Review by JamieMagee — DISMISSED** [review]

File: lib/modules/versioning/semver-coerced/index.ts [hunk 4]
```diff
@@ -77,9 +96,10 @@ function isLessThanRange(version: string, range: string): boolean {
 function isGreaterThan(version: string, other: string): boolean {
   const coercedVersion = semver.coerce(version);
   const coercedOther = semver.coerce(other);
-  return coercedVersion && coercedOther
-    ? semver.gt(coercedVersion, coercedOther)
-    : false;
+  if (!coercedVersion || !coercedOther) {
+    return false;
+  }
+  return semver.gt(coercedVersion, coercedOther);
```
Comment (viceice) [inline lib/modules/versioning/semver-coerced/index.ts:102]: why is this changed?

**Review by JamieMagee — APPROVED** [review]

File: lib/modules/versioning/semver-coerced/index.ts (hunk 4 above)
Comment (nabeelsaabna) [reply lib/modules/versioning/semver-coerced/index.ts:102]: nothing but to make the `return false;` more clear, I don't mind reverting it if you wish so.

**Review by viceice — APPROVED** [review]

**Pull request merged**